src/main/scala/mesosphere/mesos/EnvironmentHelper.scala

```diff
@@ -0,0 +1,105 @@
+package mesosphere.mesos
+
+import mesosphere.marathon.state.AppDefinition
+
+import scala.collection.immutable.Seq
+import scala.util.Random
+
+object EnvironmentHelper {
+  val maxEnvironmentVarLength = 512
+  val labelEnvironmentKeyPrefix = "MARATHON_APP_LABEL_"
+  val maxVariableLength = maxEnvironmentVarLength - labelEnvironmentKeyPrefix.length
+
+  // portsEnv generates $PORT{x} and $PORT_{y} environment variables, wherein `x` is an index into
+  // the portDefinitions or portMappings array and `y` is a non-zero port specifically requested by
+  // the application specification.
+  //
+  // @param requestedPorts are either declared container ports (if port mappings are specified) or host ports;
+  // may be 0's
+  // @param effectivePorts resolved non-dynamic host ports allocated from Mesos resource offers
+  // @return a dictionary of variables that should be added to a tasks environment
+  //scalastyle:off cyclomatic.complexity method.length
+  def portsEnv(
+    requestedPorts: Seq[Int],
+    effectivePorts: Seq[Option[Int]],
+    portNames: Seq[Option[String]]): Map[String, String] = {
+    if (effectivePorts.isEmpty) {
+      Map.empty
+    } else {
+      val env = Map.newBuilder[String, String]
+      val generatedPortsBuilder = Map.newBuilder[Int, Int] // index -> container port
+
+      object ContainerPortGenerator {
+        // track which port numbers are already referenced by PORT_xxx envvars
+        lazy val consumedPorts = scala.collection.mutable.Set(requestedPorts: _*) ++= effectivePorts.flatten
+        val maxPort: Int = 65535 - 1024
+
+        // carefully pick a container port that doesn't overlap with other ports used by this
+        // container. and avoid ports in the range (0 - 1024)
+        def next: Int = {
+          val p = Random.nextInt(maxPort) + 1025
+          if (!consumedPorts.contains(p)) {
+            consumedPorts += p
+            p
+          } else next // TODO(jdef) **highly** unlikely, but still possible that the port range could be exhausted
```

Comment:
this is a refactoring of what exists already on master. not unique per-cluster/agent. container ports are only used in the context of a container's network namespace. more like .. unique per runnableSpec